Updated the native brotli library to the latest version.

Fraser999 opened the pull request:
The latest brotli encoding lib is now a C library rather than C++, so this has allowed removal
of the `brotli_capi` wrappers.

The main motivation for this PR is to enable brotli2-rs to be cross-compiled for musl easily.  Since musl doesn't provide a g++ wrapper, the inclusion of C++ files in the compression library is difficult to work around.

I ported the remaining small amount of C++ code in [`BrotliCompress()`](https://github.com/google/brotli/blob/master/enc/compressor.cc#L86-L136), [`BrotliMemIn`](https://github.com/google/brotli/blob/master/enc/streams.cc#L58-L80) and [`MyBrotliOut`](https://github.com/Fraser999/brotli2-rs/blob/master/brotli-sys/src/brotli_capi.cc#L82-L95) to Rust in [`compress_vec()`](https://github.com/Fraser999/brotli2-rs/blob/use_c_lib/src/stream.rs#L391-L445).

alexcrichton commented:
Whoa, nice! I always though it was kinda odd they were still using C++.

Looks like there's some test failures but they're a little obscure, I'll take a look

alexcrichton commented on brotli-sys/build.rs:
```diff
@@ -16,33 +16,29 @@ fn main() {
-        .file("brotli/enc/static_dict.cc")
-        .file("brotli/enc/streams.cc")
-        .file("brotli/enc/utf8_util.cc")
+        .file("brotli/enc/backward_references.c")
```
Can both this and the decoder be compiled at the same time now? I vaguely remember that before they hit errors one way or another, but it'd be great to compile them all in one shot.

Fraser999 replied:
Yup - looks like they can.

We do get multiple definition errors for `BROTLI_ALLOC` and `BROTLI_FREE` if we `#include` both [dec/port.h](https://github.com/google/brotli/blob/master/dec/port.h#L152-L157) and [enc/memory.h](https://github.com/google/brotli/blob/master/enc/memory.h#L42-L48) which will probably require skipping more checks in the systest unfortunately.

alexcrichton commented on brotli-sys/src/lib.rs:
```diff
@@ -206,71 +201,256 @@ extern {
+pub const RUST_OPERATION_FINISH: BrotliEncoderOperation = 2;
+
+pub const RUST_DEFAULT_QUALITY: u32 = 11;
+pub const RUST_DEFAULT_WINDOW: u32 = 22;
```
I think all the `RUST_*` constants no longer exist now, right?

Fraser999 replied:
Yes, these should be `BROTLI_*`.

alexcrichton commented on brotli-sys/src/lib.rs:
```diff
@@ -206,71 +201,256 @@ extern {
+
+#[repr(C)]
+pub struct H2 {
+    pub buckets_: [u32; (1 << 16) + 1],
```
Unfortunately the automated tests for the C bindings I think are choking on this, perhaps the size could be directly inlined here (e.g. calculate `(1 << 16) + 1`)

Fraser999 replied:
Will do.

alexcrichton commented on src/stream.rs:
```diff
@@ -196,8 +204,22 @@ impl Compress {
+                                                  params.lgwin);
+            brotli_sys::BrotliEncoderSetParameter(state,
+                                                  brotli_sys::RUST_PARAM_LGBLOCK,
+                                                  params.lgblock);
```
It looks like the encoder can be created before setting the parameters, so the `CompressParams` struct may no longer make sense. Perhaps this could remove `CompressParams` altogether and move the various configuration methods to configuration on `Compress` directly?

alexcrichton replied:
Oh wait nevermind, there's still `compress_buf` below.

In that case perhaps `new` could take no arguments and there could be a `set_params` method?

Fraser999 replied:
I could do that if you prefer, but I can't see the win to be honest.  ~~Won't most users avoid the `Compress` struct altogether and go straight for the `compress_buf` or `compress_vec` functions?~~

alexcrichton commented on src/stream.rs:
```diff
@@ -25,29 +27,35 @@ unsafe impl Sync for Decompress {}
+    pub lgwin: u32,
+    /// Base 2 logarithm of the maximum input block size. Range is 16 to 24. If set to 0, the value
+    /// will be set based on the quality.
+    pub lgblock: u32,
```
I personally prefer to leave these as private and allow mutation through the builder-style interface. That way we can add fields later on without breaking backwards compatibility

Fraser999 replied:
Sure thing.

alexcrichton commented on src/stream.rs:
```diff
@@ -439,18 +483,15 @@ impl CompressParams {
+            CompressMode::Text => brotli_sys::RUST_MODE_TEXT,
+            CompressMode::Font => brotli_sys::RUST_MODE_FONT,
         }
     }
```
It should be ok to drop `unsafe` from this method I believe

Fraser999 replied:
Good point.

alexcrichton commented on src/stream.rs:
```diff
@@ -367,47 +391,73 @@ pub fn compress_buf(params: &CompressParams,
 pub fn compress_vec(params: &CompressParams,
```
This was actually largely only provided because there was a raw binding in the underlying library, but this is actually quite interesting now! It looks like `BrotliEncoderCompressStream` is more low level than `BrotliEncoderWriteData` used above in `Compress`, perhaps that should switch to using this API instead?

Fraser999 replied:
Yes - I'll have a look.

Fraser999 replied:
Actually, it seems like it's the other way round.  `BrotliEncoderCompressStream` internally calls `BrotliEncoderWriteData`.

I guess if anything, I should change `compress_vec` to use the `Compress` functions.  That should make it much simpler and will allow the removal of another `unsafe` block.

alexcrichton replied:
Ah ok, make sense to just bottom out at the "most low level" API as soon as possible for us I think. I'd be fine with just dropping this function entirely as it's somewhat niche anyway.

Fraser999 replied:
:D  That's the only compression function we use!  It's fine to remove this though - I'll update our code to just use the `write::BrotliEncoder`.

Fraser999 commented:
I've updated this now.  Points to note:

- I haven't changed `Compress::new()`, but can do if you wish.
- I've put a kind of dirty workaround in for the multiple definitions error we get in the systest.  You could maybe suggest a better approach?
- I've unfortunately had to skip systests for two of the encoder structs (one being the main state holder) since their implementations aren't public.  Again, if there's a way I can avoid this, I'd be happy to make the changes.
- I've simplified the `compress_vec()` function by using `Compress::copy_input()` and `Compress::compress()`, since this is actually making use of the lowest level API of brotli.  I can change this back and/or look at modifying the `Compress` interface to look more like `BrotliEncoderCompressStream`, but I'm not convinced that either of these things would be an improvement.

alexcrichton commented on src/stream.rs:
```diff
@@ -196,8 +203,22 @@ impl Compress {
+                                                  params.lgwin);
+            brotli_sys::BrotliEncoderSetParameter(state,
+                                                  brotli_sys::BROTLI_PARAM_LGBLOCK,
+                                                  params.lgblock);
```
Actually if we remove `compress_vec` below then we can jettison `CompressParams` entirely and just make these methods on the `Compress` structure itself. I'll just do that locally though, the rest of this looks good to go.

alexcrichton commented on systest/build.rs:
```diff
@@ -3,25 +3,37 @@ extern crate ctest;
+       .header("encode.h")
+       .header("hash.h")
+       .header("ringbuffer.h")
+       .header("workaround.h");  // this #includes state.h, first undefining BROTLI_ALLOC and BROTLI_FREE to avoid multiple definition warnings.
```
That's actually pretty hilarious, but totally fine for this :)

alexcrichton commented on systest/build.rs:
```diff
@@ -3,25 +3,37 @@ extern crate ctest;
+    });
+    cfg.skip_type(|n| n == "__enum_ty" || n == "BrotliEncoderState" || n == "BrotliEncoderStreamState");
     cfg.skip_signededness(|s| s.ends_with("_func") || s == "BrotliState");
+    cfg.skip_struct(|s| s == "BrotliEncoderStateStruct");
```
Hm ideally we wouldn't exempt these, but I'll investigate locally.

alexcrichton commented:
I've got one or two more cleanups in mind I'll do after merging, but this all looks great to me, thanks @Fraser999!

alexcrichton commented:
Ok, a few more commits pushed to master, could you give that a spin and see if it works for you? If so I'll make a new release.

Fraser999 commented:
I added a couple of comments to your other commits, but basically this is working fine for us thanks.

alexcrichton commented:
Cool, thanks @Fraser999! I've updated to 0.2.0 as well and published to crates.io

Fraser999 commented:
Nice one :)  Thanks.
